libgalois/src/GraphTopology.cpp

```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+    // set all edges incident to projected nodes
+    katana::do_all(
+        katana::iterate(Node{0}, Node{num_new_nodes}),
+        [&](auto src) {
+          auto old_src = new_to_old_nodes_mapping[src];
+          for (Edge e : topology.edges(old_src)) {
+            auto dest = topology.edge_dest(e);
+            if (bitset_nodes.test(dest)) {
+              bitset_edges.set(e);
+              out_indices[src] += 1;
+            }
+          }
+        },
+        katana::steal());
+  } else {
+    std::set<katana::EntityTypeID> edge_entity_type_ids;
+
+    for (auto edge_type : edge_types) {
+      auto entity_type_id = pg->GetEdgeEntityTypeID(edge_type);
+      edge_entity_type_ids.insert(entity_type_id);
+    }
+
+    katana::GAccumulator<uint32_t> accum_num_new_edges;
+
+    katana::do_all(
+        katana::iterate(Node{0}, Node{num_new_nodes}),
+        [&](auto src) {
+          auto old_src = new_to_old_nodes_mapping[src];
+
+          for (Edge e : topology.edges(old_src)) {
+            auto dest = topology.edge_dest(e);
+            if (bitset_nodes.test(dest)) {
+              for (auto type : edge_entity_type_ids) {
+                if (pg->DoesEdgeHaveType(e, type)) {
+                  accum_num_new_edges += 1;
+                  bitset_edges.set(e);
+                  out_indices[src] += 1;
+                  break;
+                }
+              }
+            }
+          }
+        },
+        katana::steal());
+
+    num_new_edges = accum_num_new_edges.reduce();
+  }
+
+  // Prefix sum calculation of the edge index array
+  katana::ParallelSTL::partial_sum(
+      out_indices.begin(), out_indices.end(), out_indices.begin());
+
+  NUMAArray<Edge> out_dests_offset;
+  out_dests_offset.allocateInterleaved(num_new_nodes);
+
+  // temporary buffer for storing the starting point of each node's
+  // adjacency
+  out_dests_offset[0] = 0;
+  katana::do_all(
+      katana::iterate(Node{1}, Node{num_new_nodes}),
+      [&](Node n) { out_dests_offset[n] = out_indices[n - 1]; },
+      katana::no_stats());
+
+  NUMAArray<Node> out_dests;
+  NUMAArray<Edge> old_to_new_edges_mapping;
+  NUMAArray<Edge> new_to_old_edges_mapping;
+
+  out_dests.allocateInterleaved(num_new_edges);
+  old_to_new_edges_mapping.allocateInterleaved(topology.num_edges());
+  new_to_old_edges_mapping.allocateInterleaved(num_new_edges);
+
+  // Update out_dests with the new destination ids
+  katana::do_all(
+      katana::iterate(Node{0}, Node{num_new_nodes}),
+      [&](Node n) {
+        auto src = new_to_old_nodes_mapping[n];
+
+        for (Edge e : topology.edges(src)) {
+          if (bitset_edges.test(e)) {
+            auto e_new = out_dests_offset[n];
+            out_dests_offset[n]++;
+
+            auto dest = topology.edge_dest(e);
+            dest = old_to_new_nodes_mapping[dest];
+            out_dests[e_new] = dest;
+
+            old_to_new_edges_mapping[e] = e_new;
+            new_to_old_edges_mapping[e_new] = e;
```

Comment:
Please, comment here about the invariants that help you avoid a race condition here.

Reply:
again there is no race condition here. locations that are updated are only written by one specific thread.